### PR TITLE
Optimize slow bottleneck tests of TestDaemonStartWithDaemonCommand

### PR DESCRIPTION
**Pull request by wenchma**
- Remvoe integration test of TestDaemonStartWithDaemonCommand
- Rewrite as unit test

Related issue #19425 

Before:

```
DockerSuite.TestDaemonStartWithDaemonCommand                    25.648s
```

Test results After:

```
PASS: docker_cli_daemon_test.go:434: DockerSuite.TestDaemonStartWithDaemonCommand   2.267s
```

Signed-off-by: Wen Cheng Ma wenchma@cn.ibm.com

### CODE REVIEW

File: integration-cli/docker_cli_daemon_test.go [hunk 1]
```diff
@@ -466,17 +466,6 @@ func (s *DockerSuite) TestDaemonStartWithDaemonCommand(c *check.C) {
...
-	// should error out
-	for _, f := range flags {
-		d := NewDaemon(c)
-		d.GlobalFlags = f[common]
-		if err := d.Start(f[daemon]...); err == nil {
-			d.Stop()
-			c.Fatalf("Daemon should have failed to start with docker %v daemon %v", d.GlobalFlags, f[daemon])
-		}
-	}
-
```
Comment (wenchma) [inline integration-cli/docker_cli_daemon_test.go]: these tests pass `daemon` flags as global flags,  duplicated with the following tests.

**Comment by wenchma** [comment]
@tiborvass PTAL

**Comment by wenchma** [comment]
looks like that the failed jenkins checks are not related with this PR

**Comment by tiborvass** [comment]
@wenchma thanks, i think this whole test should be deleted and rewritten as a unit test in `docker/` folder.

**Comment by wenchma** [comment]
@tiborvass thanks for your suggestion, I have updated this PR, PTAL again !

**Comment by calavera** [comment]
LGTM

File: integration-cli/docker_cli_daemon_test.go [hunk 2]
```diff
@@ -432,69 +432,6 @@ func (s *DockerDaemonSuite) TestDaemonLogLevelWrong(c *check.C) {
...
-	type kind int
-
-	const (
-		common kind = iota
-		daemon
-	)
-
-	var flags = []map[kind][]string{
-		{common: {"-l", "info"}, daemon: {"--selinux-enabled"}},
-		{common: {"-D"}, daemon: {"--selinux-enabled", "-r"}},
```
Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: @wenchma I don't see `-D` in the new unit tests is that on purpose?

File: docker/daemon_test.go [hunk 3]
```diff
@@ -34,6 +34,49 @@ func TestLoadDaemonCliConfigWithoutOverriding(t *testing.T) {
 	}
 }
 
+func TestLoadDaemonCliConfigWithDaemonFlags(t *testing.T) {
```
Comment (tiborvass) [inline docker/daemon_test.go]: this is does not seem to be the same test as TestDaemonStartWithDaemonCommand. It seems this unit test is testing the config file, whereas TestDaemonStartWithDaemonCommand is testing flags on the daemon cli.

File: integration-cli/docker_cli_daemon_test.go (hunk 2 above)
Comment (wenchma) [inline integration-cli/docker_cli_daemon_test.go]: @tiborvass will add '-D' cover.

File: docker/daemon_test.go (hunk 3 above)
Comment (wenchma) [inline docker/daemon_test.go]: @tiborvass the function  `loadDaemonCliConfig`  requires daemon flags and also config file, so this unit test includes the daemon flags, and also config file,  we can leave config file empty.  right ?

**Comment by estesp** [comment]
I think this looks good for the valid flag coverage (@tiborvass do you have any more issue given the responses above?).  However, I wonder if we need to keep the invalid global flag piece as a integration-test but use something like the goroutines that have been added to other "slow" integration tests for failure cases--you could run all 3 "bad flag" `NewDaemon` tests in parallel and then make sure you have errors for all 3.  See #20914 for an example.

**Comment by tiborvass** [comment]
Nvm it's fine like this, I just realized the actual cli code in docker could be further refactored, since we completely deprecated `--daemon`, so we don't need to do that stupid dance with global daemon flags.

LGTM

**Pull request merged**